### PR TITLE
[Index] Always skip invalid locations outside of modules

### PR DESCRIPTION
**Pull request by bnbarham**
There's a number of paths that end up calling into `initIndexSymbol`
that may or may not have checked that the given location is actually
valid. Add an extra check within the method to guarantee we only output
0:0 locations when indexing modules.

Resolves rdar://98645486.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
I've seen records with 0:0 for non-module files where the only roles are Ref|RelCont. I can't see a path where that could be the case, but I must be missing one 🤷.

File: lib/Index/Index.cpp [hunk 1]
```diff
@@ -1651,17 +1646,14 @@ bool IndexSwiftASTWalker::initIndexSymbol(ExtensionDecl *ExtD, ValueDecl *Extend
...
-  std::tie(Info.line, Info.column, Info.offset) = getLineColAndOffset(Loc);
+  std::tie(Info.line, Info.column) = getLineCol(Loc);
   if (auto Group = ExtD->getGroupName())
     Info.group = Group.getValue();
   return false;
 }
 
 bool IndexSwiftASTWalker::initFuncDeclIndexSymbol(FuncDecl *D,
                                                   IndexSymbol &Info) {
-  if (initIndexSymbol(D, D->getLoc(/*SerializedOK*/false), /*IsRef=*/false, Info))
```
Comment (benlangmuir) [inline lib/Index/Index.cpp:1662]: What's the reason for this change? I find it confusing that `initFuncDeclIndexSymbol` does not imply `initIndexSymbol`.  There are several other `init<thing>IndexSymbol` methods and they all call `initIndexSymbol`.  I think we should keep them consistent, and it feels kind of unrelated to the rest of the PR.

File: lib/Index/Index.cpp [hunk 2]
```diff
@@ -1601,8 +1601,15 @@ bool IndexSwiftASTWalker::reportImplicitConformance(ValueDecl *witness, ValueDec
 bool IndexSwiftASTWalker::initIndexSymbol(ValueDecl *D, SourceLoc Loc,
                                           bool IsRef, IndexSymbol &Info) {
   assert(D);
+
+  // This *should* have been checked already, but we've definitely seen for
```
Comment (benlangmuir) [inline lib/Index/Index.cpp:1605]: Should we assert as well? That may help shake out where this is coming from.

File: lib/Index/Index.cpp (hunk 1 above)
Comment (bnbarham) [reply lib/Index/Index.cpp:1662]: Personally I found this confusing 😅. I'd prefer `initIndexSymbol` be called explicitly and then the other "inits" be "addAdditional...". The 2 other `init*IndexSymbol` are both for refs, which I didn't change because one is used in a couple places (which I guess is a good reason to keep this).

Anyway, I don't feel particularly strongly about this so if you do I'm happy to change back.

FWIW it's "related" because I was going through all these calls when trying to find paths. It's a separate commit, happy to move to another PR (or, again, change back).

File: lib/Index/Index.cpp (hunk 2 above)
Comment (bnbarham) [reply lib/Index/Index.cpp:1605]: I wasn't sure about this. When I was looking through the calls, almost all of them are just a check on the location before calling this method. There's a couple times where we'd avoid a small amount of work but IMO it would be better to do that and have the checks in the one location. Everything already checks the result of `initIndexSymbol` anyway.

So I *almost* did that, but thought it would be better to get this up first.

If you disagree with that then I can add this assert in, though there's a few places that need the location check before doing that (they aren't the cause of the 0:0 I've seen because they all add additional roles, but they could maybe be hit).

Comment (bnbarham) [reply lib/Index/Index.cpp]: Okay, I've just changed the comment for now 👍

File: lib/Index/Index.cpp (hunk 1 above)
Comment (bnbarham) [reply lib/Index/Index.cpp]: Changed this back but left the rest 👍

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by benlangmuir — APPROVED** [review]

**Review by hamishknight — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test Linux platform

**Comment by bnbarham** [comment]
@swift-ci please test Linux platform
